website/blog/2022-08-12-model-timing.md

```diff
@@ -0,0 +1,294 @@
+
+If you’re curious, here’s a look at the query for `dbt_model_summary` before any changes were made:
+
+```sql
+{{config(materialized = 'ephemeral')}}
+
+with model_execution as (
+
+    select * from {{ ref('stg_dbt_run_model_events') }}
+
+),
+
+diffed as (
+
+    select *,
+    
+        row_number() over (
+            partition by project_id, model_id
+            order by dvce_created_tstamp
+        ) = 1 as is_new,
+
+        /*
+            The `mode` window function returns the most common content hash for a
+            given model on a given day. We use this a proxy for the 'production'
+            version of the model, running in deployment. When a different hash 
+            is run, it likely reflects that the model is undergoing development.
+        */
+        
+        contents != mode(contents) over (
+            partition by project_id, model_id, dvce_created_tstamp::date
+        ) as is_changed
+        
+    from model_execution
+    
+),
+
+final as (
+
+    select
+        invocation_id,
+        max(model_complexity) as model_complexity,
+        max(model_total) as count_models,
+        sum(case when is_new or is_changed then 1 else 0 end) as count_changed,
+        sum(case when skipped = true then 1 else 0 end) as count_skip,
+        sum(case when error is null or error = 'false' then 0 else 1 end) as count_error,
+        sum(case when (error is null or error = 'false') and skipped = false then 1 else 0 end) as count_succeed
+
+    from diffed
+    group by 1
+
+)
+
+select * from final
+```
+
+The window functions referenced above are answering the following questions:
+
+- `row_number()`
+    - *Is this the first time that this specific model has run in a project?*
+    - Note: this grain is at the project level
+- `mode()`
+    - *Is this the most frequent version of the model that ran today (based on the hashed contents)?*
+    - Note: this grain is at the model + run date level
+
+## Moving to Solutions
+
+### Attempt #1: Optimizing our objects and materializations
+
+Given the size and complexity of this query, the first few approaches we took didn’t focus on changing the query as much as optimizing our objects and materializations. 
+
+The two window functions (`row_number()` and `mode()` in the `diffed` CTE above) were in an [ephemeral model](https://docs.getdbt.com/docs/building-a-dbt-project/building-models/materializations#ephemeral) which isn’t stored in the data warehouse, but is instead executed in-memory at run time. Since it was obvious our virtual warehouse was running out of memory (remote storage spillage), we tried swapping that to a view, then a table materialization. Neither of these improved the run time significantly, so we tried clustering the table. However, since our two window functions are at different grains there wasn’t a great clustering key we found for this.
```

Comment:
```suggestion
The two window functions (`row_number()` and `mode()` in the `diffed` <Term id="cte" /> above) were in an [ephemeral model](https://docs.getdbt.com/docs/building-a-dbt-project/building-models/materializations#ephemeral) which isn’t stored in the data warehouse, but is instead executed in-memory at run time. Since it was obvious our virtual warehouse was running out of memory (remote storage spillage), we tried swapping that to a view, then a table materialization. Neither of these improved the run time significantly, so we tried clustering the table. However, since our two window functions are at different grains there wasn’t a great clustering key we found for this.
```